### PR TITLE
Added Tests for Gem Catalog Filtering

### PR DESCRIPTION
**Pull request by AMZN-nggieber**
No new bugs were found through the development of tests (besides display name being set) but it did bring to question whether the intended results from some of the filters are the most desirable. Most filters work on an or basis relative to themselves and an and basis relative to other filters. 

This is outside of the scope of this PR but it might be worth considering whether a user would prefer platform selection to work on an or basis or (as it does currently) or an and basis.Using an or basis if a user selected the Windows, Android and iOS platforms all gems that supported at least one of those platforms would be shown. Instead the user might prefer it to work on an and basis wherein only the gems that supported all of those platforms would be shown. This way the user could determine their targeted platforms and select them making it easy to see the full set of gems that are compatible with all of their target platforms.

`[----------] Global test environment tear-down
[==========] 36 tests from 7 test cases ran. (5337 ms total)
[  PASSED  ] 36 tests.`

Signed-off-by: nggieber <52797929+AMZN-nggieber@users.noreply.github.com>

### CODE REVIEW

File: Code/Tools/ProjectManager/tests/GemCatalogTests.cpp [hunk 1]
```diff
@@ -35,30 +43,530 @@ namespace O3DE::ProjectManager
...
+
+        QVector<int> m_gemRows;
+    };
+
+    TEST_F(GemCatalogSearchFilterTests, GemCatalog_Filters_By_Search_String_Name_Only)
+    {
+        m_proxyModel->SetSearchString("Name");
+
+        EXPECT_TRUE(m_proxyModel->filterAcceptsRow(m_gemRows[Name], QModelIndex()));
+        EXPECT_TRUE(m_proxyModel->filterAcceptsRow(m_gemRows[DisplayName], QModelIndex()));
```
Comment (AMZN-alexpete) [inline Code/Tools/ProjectManager/tests/GemCatalogTests.cpp:134]: Interesting that the "Name" only searches the "Name" and "DisplayName", but the "Display Name" only, just searches the "DisplayName" and not "Name"  - Is this intentional?

File: Code/Tools/ProjectManager/tests/GemCatalogTests.cpp [hunk 2]
```diff
@@ -35,30 +43,530 @@ namespace O3DE::ProjectManager
...
+        m_proxyModel->SetPlatforms(GemInfo::Platform::macOS);
+
+        EXPECT_FALSE(m_proxyModel->filterAcceptsRow(m_gemRows[DefaultAudio], QModelIndex()));
+        EXPECT_FALSE(m_proxyModel->filterAcceptsRow(m_gemRows[MobileUX], QModelIndex()));
+        EXPECT_TRUE(m_proxyModel->filterAcceptsRow(m_gemRows[CityProps], QModelIndex()));
+    }
+
+    TEST_F(GemCatalogMiscFilterTests, GemCatalog_Filters_By_Platform_Multiselection_Shows_All_Matches)
+    {
+        m_proxyModel->SetPlatforms(GemInfo::Platform::Windows | GemInfo::Platform::iOS);
```
Comment (Kadino) [inline Code/Tools/ProjectManager/tests/GemCatalogTests.cpp:527]: This sets up a different filtering scenario than the other multiselects, in that all three elements are found. The negative filtering case isn't being verified, due to how the platform labels line up.

I don't see a good filter that causes two different groups singly match, so perhaps modify the registered platform info to make sure one row gets excluded by the filter?

File: Code/Tools/ProjectManager/tests/GemCatalogTests.cpp (hunk 1 above)
Comment (AMZN-nggieber) [reply Code/Tools/ProjectManager/tests/GemCatalogTests.cpp:134]: Yes this is purposeful. Token matching is not supported and is specifically called about by a later test below called `GemCatalog_Filters_By_Search_String_No_Token_Support`

Not to say that it won't likely be supported in the future, it just isn't right now.

**Review by Kadino — COMMENTED** [review]
Some minor suggestions:

File: Code/Tools/ProjectManager/tests/GemCatalogTests.cpp [hunk 3]
```diff
@@ -35,30 +43,530 @@ namespace O3DE::ProjectManager
...
+            DisplayName,
+            Creator,
+            Summary,
+            Features
+        };
+
+        QVector<int> m_gemRows;
+    };
+
+    TEST_F(GemCatalogSearchFilterTests, GemCatalog_Filters_By_Search_String_Name_Only)
```
[the remark this answers is not in the document]
Comment (AMZN-nggieber) [reply Code/Tools/ProjectManager/tests/GemCatalogTests.cpp:129]: Thanks for the link I look through this and rethink all the names!

File: Code/Tools/ProjectManager/tests/GemCatalogTests.cpp [hunk 4]
```diff
@@ -35,30 +43,530 @@ namespace O3DE::ProjectManager
...
+        GemModel::SetIsAdded(*m_gemModel, indexA, true);
 
         // expect b and c are now dependencies of an added gem but not themselves added
         // cmake will handle dependencies
         EXPECT_TRUE(GemModel::IsAddedDependency(indexB) && GemModel::IsAddedDependency(indexC));
-        EXPECT_TRUE(!GemModel::IsAdded(indexB) && !GemModel::IsAdded(indexC));
+        EXPECT_FALSE(GemModel::IsAdded(indexB) || GemModel::IsAdded(indexC));
 
-        QVector<QModelIndex> gemsToAdd = gemModel->GatherGemsToBeAdded();
+        QVector<QModelIndex> gemsToAdd = m_gemModel->GatherGemsToBeAdded();
```
Comment (AMZN-alexpete) [inline Code/Tools/ProjectManager/tests/GemCatalogTests.cpp:63]: can we const & this?

**Review by AMZN-alexpete — APPROVED** [review]

**Review by AMZN-ScottR — APPROVED** [review]
Agree with @Kadino on the test names, otherwise looks good to me.

**Pull request merged**